Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs

```diff
@@ -165,25 +160,26 @@ void OnDisconnectOrTimeout(ulong clientID)
 
                 //On a client disconnect we want to take them back to the main menu.
                 //We have to check here in SceneManager if our active scene is the main menu, as if it is, it means we timed out rather than a raw disconnect;
-                if (SceneManager.GetActiveScene().name != "MainMenu")
+                switch (DisconnectReason.Reason)
                 {
-                    if (DisconnectReason.Reason == ConnectStatus.UserRequestedDisconnect || DisconnectReason.Reason == ConnectStatus.HostDisconnected || NetworkManager.Singleton.IsHost)
-                    {
-                        // simply shut down and go back to main menu
-                        m_ApplicationController.LeaveSession();
-                    }
-                    else
-                    {
+                    case ConnectStatus.UserRequestedDisconnect:
+                    case ConnectStatus.HostDisconnected:
+                    case ConnectStatus.LoggedInAgain:
+                    case ConnectStatus.ServerFull:
+                        m_ApplicationController.LeaveSession(); // go through the normal leave flow
+                        break;
+                    case ConnectStatus.GenericDisconnect:
+                    case ConnectStatus.Undefined:
                         DisconnectReason.SetDisconnectReason(ConnectStatus.Reconnecting);
+
                         // load new scene to workaround MTT-2684
                         SceneManager.LoadScene("Loading");
+
                         // try reconnecting
                         m_TryToReconnectCoroutine ??= StartCoroutine(TryToReconnect(lobbyCode));
```

Comment:
that's weird, in OnConnectFinished, we stop the coroutine when succesfully reconnecting. I only tested by timing out a connected client by pausing, but I don't see why that would work differently with separate machines?

Also the original reason I was not re-initiating the coroutine from scratch is to handle some edge cases like if managing to reconnect before the timeout occurs server-side as well and our connection gets denied because we were technically considered a duplicate connection. In that case, we would simply do nothing and let the reconnect coroutine do a second reconnect attempt (although now we only have one attempt so if we don't plan on increasing it back sure we can remove some complexity from this)

Reply:
Yeah the way I was able to reproduce this is:

1. Machine A host
2. Machine B join
3. B drop connection, reconnect
4. Reconnect works for B (noting that I added more reconnect attempts and on the logs I reconnected on attempt 3/5)
5. Repeat 3
6. Reconnect for B will pick up at attempt 4